Implement adding bias vector into structured sparse linear operator

alexsamardzic opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100881



cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer @albanD

Differential Revision: [D46453477](https://our.internmc.facebook.com/intern/diff/D46453477)

alexsamardzic committed:
Implement adding bias vector into structured sparse linear operator

[ghstack-poisoned]

alexsamardzic commented:
<details>
<summary>Benchmark script for comparing `_structured_sparse_linear()` performance with `nn.functional.linear()`</summary>

```python
#! /usr/bin/env python

import torch
import torch.utils.benchmark as benchmark
from torch import nn
from torch.ao.pruning import WeightNormSparsifier
from itertools import product

device = 'cuda'
dtype = torch.float16
torch.set_printoptions(precision=3, threshold=None, edgeitems=4, linewidth=460, profile=None, sci_mode=False)

if __name__ == '__main__':
    results = []

    shapes = [
        # distilbert shapes
        (768, 3072, 768),
        (3072, 768, 3072),
        # jiecao shapes
        (1024, 1536, 2048),
        (1024, 9408, 2048),
        (1024, 3200, 2048),
        (1024, 256, 9472),
        (1024, 10240, 256),
        (1024, 256, 12608),
        (1024, 2560, 1024),
        (1024, 512, 10240),
        (1024, 10240, 512),
        (1024, 2048, 1024),
        (1024, 512, 512),
        (1024, 1024, 1024),
        (1024, 2048, 2048),
        (2048, 1536, 2048),
        (2048, 9408, 2048),
        (2048, 3200, 2048),
        (2048, 256, 9472),
        (2048, 10240, 256),
        (2048, 256, 12608),
        (2048, 2560, 1024),
        (2048, 512, 10240),
        (2048, 10240, 512),
        (2048, 2048, 1024),
        (2048, 512, 512),
        (2048, 1024, 1024),
        (2048, 2048, 2048),
    ]
    batch_sizes = [4, 16, 64, 256]


    for (m, k, n), batch_size in product(shapes, batch_sizes):
        print(m, k, n, batch_size)
        # label and sub_label are the rows
        # description is the column
        try:
            label = 'linear vs. _structured_sparse_linear'
            sub_label = f'm:{m:5d} | k:{k:5d} | n:{n:5d} | batch_size: {batch_size:4d}'

            model = nn.Linear(k, m).to(dtype).to(device)
            model.eval()

            pruner = WeightNormSparsifier(sparsity_level=1.0, sparse_block_shape=(1, 4), zeros_per_block=2)
            pruner.prepare(model, [{'tensor_fqn': 'weight'}])
            pruner.step()
            pruner.squash_mask()

            model.weight.required_grad = False
            model.bias.required_grad = False

            dense_weight = model.weight
            mask = dense_weight != 0
            sparse_weight = dense_weight.masked_select(mask).view(m, k // 2)
            bias = model.bias

            for i in range(2):
                input = torch.randn(batch_size, n, k, device=device, dtype=dtype)
                dense_output = nn.functional.linear(input, dense_weight, bias)
                if i == 0:
                    sparse_output, meta = torch._structured_sparse_linear(input, sparse_weight, mask, bias)
                else:
                    sparse_output, _ = torch._structured_sparse_linear(input, sparse_weight, meta, bias)

                assert torch.allclose(sparse_output, dense_output, rtol=1e-3, atol=1e-3)
            
            measurement = benchmark.Timer(
                stmt='dense_linear(input, weight, bias)',
                globals={'dense_linear': nn.functional.linear, 'input': input, 'weight': dense_weight, 'bias': bias},
                label=label,
                sub_label=sub_label,
                description='dense latency',
            ).blocked_autorange()
            results.append(measurement)

            measurement = benchmark.Timer(
                stmt='sparse_linear(input, weight, mask_or_meta, bias)',
                globals={'sparse_linear': torch._structured_sparse_linear, 'input': input, 'weight': sparse_weight, 'mask_or_meta': meta, 'bias': bias},
                label=label,
                sub_label=sub_label,
                description='sparse linear latency',
            ).blocked_autorange()
            results.append(measurement)

        except Exception:
            continue

        # with profile(activities=[ProfilerActivity.CUDA, ProfilerActivity.CPU], record_shapes=True, with_stack=True) as prof:
        #     with record_function('CUTLASS'):
        #         sparse_linear(input_tensor)

        # prof.export_stacks(f'{m}_{k}_{n}_cutlass_profiler_stacks.txt', 'self_cuda_time_total')

    compare = benchmark.Compare(results)
    compare.colorize(rowwise=True)
    compare.print()
```
</details>

<details>
<summary>Benchmarking results on A100</summary>

```
[----------------------------- linear vs. structure_sparse_linear -----------------------------]
                                                      |  dense latency  |  sparse linear latency
1 threads: -------------------------------------------------------------------------------------
      m:  768 | k: 3072 | n:  768 | batch_size:    4  |        74.6     |            62.7       
      m:  768 | k: 3072 | n:  768 | batch_size:   16  |       257.9     |           190.8       
      m:  768 | k: 3072 | n:  768 | batch_size:   64  |       980.7     |           685.4       
      m:  768 | k: 3072 | n:  768 | batch_size:  256  |      3867.9     |          2696.8       
      m: 3072 | k:  768 | n: 3072 | batch_size:    4  |       265.5     |           210.1       
      m: 3072 | k:  768 | n: 3072 | batch_size:   16  |      1008.6     |           797.1       
      m: 3072 | k:  768 | n: 3072 | batch_size:   64  |      4032.7     |          3175.3       
      m: 3072 | k:  768 | n: 3072 | batch_size:  256  |     15989.8     |         12717.0       
      m: 1024 | k: 1536 | n: 2048 | batch_size:    4  |       130.6     |            89.0       
      m: 1024 | k: 1536 | n: 2048 | batch_size:   16  |       489.0     |           330.4       
      m: 1024 | k: 1536 | n: 2048 | batch_size:   64  |      1888.1     |          1293.7       
      m: 1024 | k: 1536 | n: 2048 | batch_size:  256  |      7500.6     |          5124.4       
      m: 1024 | k: 9408 | n: 2048 | batch_size:    4  |       644.1     |           484.0       
      m: 1024 | k: 9408 | n: 2048 | batch_size:   16  |      2445.9     |          1837.1       
      m: 1024 | k: 9408 | n: 2048 | batch_size:   64  |      9731.9     |          7049.0       
      m: 1024 | k: 9408 | n: 2048 | batch_size:  256  |     38844.2     |         28009.0       
      m: 1024 | k: 3200 | n: 2048 | batch_size:    4  |       242.4     |           172.3       
      m: 1024 | k: 3200 | n: 2048 | batch_size:   16  |       906.9     |           637.1       
      m: 1024 | k: 3200 | n: 2048 | batch_size:   64  |      3556.2     |          2493.7       
      m: 1024 | k: 3200 | n: 2048 | batch_size:  256  |     13665.3     |          9813.0       
      m: 1024 | k:  256 | n: 9472 | batch_size:    4  |       131.8     |           110.9       
      m: 1024 | k:  256 | n: 9472 | batch_size:   16  |       511.4     |           425.4       
      m: 1024 | k:  256 | n: 9472 | batch_size:   64  |      2009.1     |          1691.4       
      m: 1024 | k:10240 | n:  256 | batch_size:    4  |       112.4     |           114.8       
      m: 1024 | k:10240 | n:  256 | batch_size:   16  |       361.8     |           298.9       
      m: 1024 | k:10240 | n:  256 | batch_size:   64  |      1377.9     |          1020.0       
      m: 1024 | k:10240 | n:  256 | batch_size:  256  |      5348.5     |          3934.6       
      m: 1024 | k:  256 | n:12608 | batch_size:    4  |       175.4     |           146.5       
      m: 1024 | k:  256 | n:12608 | batch_size:   16  |       679.3     |           566.5       
      m: 1024 | k:  256 | n:12608 | batch_size:   64  |      2665.9     |          2263.7       
      m: 1024 | k: 2560 | n: 1024 | batch_size:    4  |       104.1     |            82.2       
      m: 1024 | k: 2560 | n: 1024 | batch_size:   16  |       398.7     |           272.9       
      m: 1024 | k: 2560 | n: 1024 | batch_size:   64  |      1471.7     |          1034.3       
      m: 1024 | k: 2560 | n: 1024 | batch_size:  256  |      5671.6     |          4083.2       
      m: 1024 | k:  512 | n:10240 | batch_size:    4  |       213.3     |           178.0       
      m: 1024 | k:  512 | n:10240 | batch_size:   16  |       825.8     |           683.5       
      m: 1024 | k:  512 | n:10240 | batch_size:   64  |      3249.7     |          2719.8       
      m: 1024 | k:10240 | n:  512 | batch_size:   16  |       696.6     |           528.4       
      m: 1024 | k:10240 | n:  512 | batch_size:   64  |      2676.3     |          1992.2       
      m: 1024 | k:10240 | n:  512 | batch_size:  256  |     10576.1     |          7794.8       
      m: 1024 | k: 2048 | n: 1024 | batch_size:    4  |        85.9     |            66.4       
      m: 1024 | k: 2048 | n: 1024 | batch_size:   16  |       328.2     |           220.5       
      m: 1024 | k: 2048 | n: 1024 | batch_size:   64  |      1201.1     |           834.3       
      m: 1024 | k: 2048 | n: 1024 | batch_size:  256  |      4599.2     |          3273.1       
      m: 1024 | k:  512 | n:  512 | batch_size:    4  |        26.7     |            17.8       
      m: 1024 | k:  512 | n:  512 | batch_size:   16  |        46.6     |            39.2       
      m: 1024 | k:  512 | n:  512 | batch_size:   64  |       172.6     |           142.9       
      m: 1024 | k:  512 | n:  512 | batch_size:  256  |       662.2     |           547.5       
      m: 1024 | k: 1024 | n: 1024 | batch_size:    4  |        50.3     |            37.8       
      m: 1024 | k: 1024 | n: 1024 | batch_size:   16  |       190.2     |           123.5       
      m: 1024 | k: 1024 | n: 1024 | batch_size:   64  |       602.8     |           464.6       
      m: 1024 | k: 1024 | n: 1024 | batch_size:  256  |      2372.4     |          1833.2       
      m: 1024 | k: 2048 | n: 2048 | batch_size:    4  |       165.8     |           113.1       
      m: 1024 | k: 2048 | n: 2048 | batch_size:   16  |       616.8     |           424.3       
      m: 1024 | k: 2048 | n: 2048 | batch_size:   64  |      2408.3     |          1677.5       
      m: 1024 | k: 2048 | n: 2048 | batch_size:  256  |      8990.0     |          6552.3       
      m: 2048 | k: 1536 | n: 2048 | batch_size:    4  |       250.2     |           171.7       
      m: 2048 | k: 1536 | n: 2048 | batch_size:   16  |       952.7     |           643.8       
      m: 2048 | k: 1536 | n: 2048 | batch_size:   64  |      3746.2     |          2523.5       
      m: 2048 | k: 1536 | n: 2048 | batch_size:  256  |     14856.1     |         10061.0       
      m: 2048 | k: 9408 | n: 2048 | batch_size:    4  |      1271.9     |           926.1       
      m: 2048 | k: 9408 | n: 2048 | batch_size:   16  |      4942.7     |          3586.3       
      m: 2048 | k: 9408 | n: 2048 | batch_size:   64  |     19566.6     |         14663.3       
      m: 2048 | k: 9408 | n: 2048 | batch_size:  256  |     78076.9     |         58347.1       
      m: 2048 | k: 3200 | n: 2048 | batch_size:    4  |       466.0     |           325.5       
      m: 2048 | k: 3200 | n: 2048 | batch_size:   16  |      1793.4     |          1252.6       
      m: 2048 | k: 3200 | n: 2048 | batch_size:   64  |      7097.0     |          4914.3       
      m: 2048 | k: 3200 | n: 2048 | batch_size:  256  |     28023.9     |         19820.6       
      m: 2048 | k:  256 | n: 9472 | batch_size:    4  |       256.5     |           213.7       
      m: 2048 | k:  256 | n: 9472 | batch_size:   16  |      1002.2     |           824.4       
      m: 2048 | k:  256 | n: 9472 | batch_size:   64  |      3966.2     |          3273.3       
      m: 2048 | k:10240 | n:  256 | batch_size:   16  |       689.8     |           521.8       
      m: 2048 | k:10240 | n:  256 | batch_size:   64  |      2641.2     |          2042.8       
      m: 2048 | k:10240 | n:  256 | batch_size:  256  |     10595.5     |          7918.4       
      m: 2048 | k:  256 | n:12608 | batch_size:    4  |       338.2     |           280.9       
      m: 2048 | k:  256 | n:12608 | batch_size:   16  |      1333.7     |          1095.9       
      m: 2048 | k:  256 | n:12608 | batch_size:   64  |      5309.1     |          4388.2       
      m: 2048 | k: 2560 | n: 1024 | batch_size:    4  |       199.5     |           139.6       
      m: 2048 | k: 2560 | n: 1024 | batch_size:   16  |       752.6     |           517.9       
      m: 2048 | k: 2560 | n: 1024 | batch_size:   64  |      2917.4     |          2054.0       
      m: 2048 | k: 2560 | n: 1024 | batch_size:  256  |     11539.9     |          7955.0       
      m: 2048 | k:  512 | n:10240 | batch_size:    4  |       415.7     |           340.7       
      m: 2048 | k:  512 | n:10240 | batch_size:   16  |      1617.6     |          1344.7       
      m: 2048 | k:  512 | n:10240 | batch_size:   64  |      6472.2     |          5353.4       
      m: 2048 | k:10240 | n:  512 | batch_size:    4  |       375.2     |           298.1       
      m: 2048 | k:10240 | n:  512 | batch_size:   16  |      1378.1     |          1027.2       
      m: 2048 | k:10240 | n:  512 | batch_size:   64  |      5355.9     |          3991.2       
      m: 2048 | k:10240 | n:  512 | batch_size:  256  |     21160.4     |         16448.4       
      m: 2048 | k: 2048 | n: 1024 | batch_size:    4  |       163.2     |           111.7       
      m: 2048 | k: 2048 | n: 1024 | batch_size:   16  |       616.6     |           419.6       
      m: 2048 | k: 2048 | n: 1024 | batch_size:   64  |      2397.4     |          1636.7       
      m: 2048 | k: 2048 | n: 1024 | batch_size:  256  |      9489.9     |          6478.9       
      m: 2048 | k:  512 | n:  512 | batch_size:    4  |        29.7     |            24.4       
      m: 2048 | k:  512 | n:  512 | batch_size:   16  |        90.3     |            74.9       
      m: 2048 | k:  512 | n:  512 | batch_size:   64  |       333.5     |           274.5       
      m: 2048 | k:  512 | n:  512 | batch_size:  256  |      1289.3     |          1062.4       
      m: 2048 | k: 1024 | n: 1024 | batch_size:    4  |        82.0     |            63.5       
      m: 2048 | k: 1024 | n: 1024 | batch_size:   16  |       305.9     |           233.9       
      m: 2048 | k: 1024 | n: 1024 | batch_size:   64  |      1164.4     |           905.2       
      m: 2048 | k: 1024 | n: 1024 | batch_size:  256  |      4699.1     |          3570.7       
      m: 2048 | k: 2048 | n: 2048 | batch_size:    4  |       312.2     |           219.4       
      m: 2048 | k: 2048 | n: 2048 | batch_size:   16  |      1191.2     |           812.6       
      m: 2048 | k: 2048 | n: 2048 | batch_size:   64  |      4781.0     |          3254.9       
      m: 2048 | k: 2048 | n: 2048 | batch_size:  256  |     18831.5     |         13150.6       

Times are in microseconds (us).
```
</details>

cpuhrsch commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3192,6 +3188,10 @@
     MkldnnCPU: mkldnn_linear_backward
   autogen: mkldnn_linear_backward.out
 
+- func: _structured_sparse_linear(Tensor input, Tensor weight, Tensor mask_or_meta, Tensor bias) -> (Tensor, Tensor)
```
Can you make bias a kwarg? It's conceivable that a user might want to run without a bias. There's a few models that do that. You can do so by using the signature
```
_structured_sparse_linear(Tensor input, Tensor weight, Tensor mask_or_meta, *, Tensor bias=None)
```
This will also fix the `backwards_compat` error.

Since this is a private function there are no backwards compatibility concerns. The way you'd resolve the `backwards_compat` CI issue is by adding the function name to [this ALLOW_LIST](https://github.com/pytorch/pytorch/blob/main/test/forward_backward_compatibility/check_forward_backward_compatibility.py#L12-L42) (see code comment above ALLOW_LIST as well).

cpuhrsch commented on aten/src/ATen/native/sparse/cuda/cutlass/custom_predicated_tile_iterator.h:
```diff
@@ -0,0 +1,519 @@
+
+#pragma once
+
+#include "cutlass/cutlass.h"
```
There are some lint errors, including the use of quotes instead of angle brackets. You can use [lintrunner](https://github.com/pytorch/pytorch/wiki/lintrunner) to run lint locally and resolve these.

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

amjames commented on aten/src/ATen/native/sparse/cuda/cutlass/README.md:
```diff
@@ -0,0 +1,29 @@
+This directory contains files from CUTLASS 3.1 source tree, modified
```
Is it possible to keep these modified CUTLASS sources temporarily and submit the modifications upstream, then remove them from the pytorch tree when a release is minted included the additional functionality? 

I worry about having to indefinitely support them breaking in subtle ways with updates to our pinned version of CUTLASS in the future. 

It would be nice to have a plan for their eventual removal.

alexsamardzic replied:
At the moment, CUTLASS supports adding bias vector to rows of a row-major output matrix, and to columns of a column-major output matrix.  As we assume that weight matrix is sparse, and as the sparse matrix has to be first operand of CTULASS sparse GEMM, this new linear operator is calculated as $(Wx^{T}+b^{T})^T$ (instead of $xW^{T}+b$, as usual); so it turns out we need to add bias vector to columns of output matrix.  In principle, we can choose any of these two layouts for the output matrix, but at the moment CUTLASS supports only row-major output for sparse GEMM.   Thus this hack.  I've opened an issue at CUTLASS GitHub, and was told creating customized PredicatedTileIterator class, outside of the CUTLASS, is the way; the rest of files are there only as, even CUTLASS is a highly templated library, it's actually cumbersome to combine at will.  So - I worry about the same thing (moreover, the fact that CUTLASS is not processing files through any kind of linter, and that we do, makes remembering what was actually changed from the original files somewhat problematic), but for now the plan will have to be to keep an eye both on updates of our pinned version, and the CUTLASS repo, to make changes here is updated of our pinned version break something, and on the other side to hope CUTLASS developers will provide some means to simplify this.

cpuhrsch commented:
@alexsamardzic - I think for now we should merge this so it's available for development, but we'll work on a follow up PR to refactor and minimize the amount of custom code.

cpuhrsch commented:
@pytorchbot merge

cpuhrsch commented:
@alexsamardzic - Can you try rebasing this PR?

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

cpuhrsch reviewed (changes requested):
It looks like you accidentally pushed a submodule update to some folders in third_party. You can resolve this by checking out the commit pin that's currently on main.

hwu36 commented:
@alexsamardzic , you could upstream cutlass code into cutlass directory and then used by torch.

alexsamardzic commented:
> @alexsamardzic , you could upstream cutlass code into cutlass directory and then used by torch.

I'd much prefer if we do it this way, but my changes of CUTLASS files are just a hack to make it work what is needed for this particular PyTorch functionality.  But if you want me, I can create a PR for CUTLASS GitHub repo, and hopefully with the help from you guys we can make it usable.

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer albanD

[ghstack-poisoned]

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer albanD

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge

cpuhrsch commented:
@pytorchbot merge

alexsamardzic commented:
> @alexsamardzic , you could upstream cutlass code into cutlass directory and then used by torch.

CUTLASS [PR 951](https://github.com/NVIDIA/cutlass/pull/951) created, let's continue discussion there.

izaitsevfb commented:
@pytorchbot revert -m "breaks internal builds, see D45972633" -c ghfirst

izaitsevfb commented:
@alexsamardzic, I'm sorry for reverting your stack, but it breaks internal builds at Meta:

```
aten/src/ATen/native/sparse/cuda/cutlass/custom_default_epilogue_tensor_op.h(347): error: namespace "cutlass" has no member "float_e4m3_t"

aten/src/ATen/native/sparse/cuda/cutlass/custom_default_epilogue_tensor_op.h(355): error: namespace "cutlass" has no member "float_e4m3_t"

aten/src/ATen/native/sparse/cuda/cutlass/custom_default_epilogue_tensor_op.h(414): error: namespace "cutlass" has no member "float_e5m2_t"

aten/src/ATen/native/sparse/cuda/cutlass/custom_default_epilogue_tensor_op.h(413): error: class template "cutlass::epilogue::threadblock::detail::CustomDefaultIteratorsTensorOp" has already been defined

aten/src/ATen/native/sparse/cuda/cutlass/custom_sparse_gemm.h(280): error: identifier "canonical_warp_idx" is undefined
```

Please ask @cpuhrsch for more context, thank you for understanding.

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer albanD

[ghstack-poisoned]

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer albanD

[ghstack-poisoned]

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer albanD

[ghstack-poisoned]

alexsamardzic commented:
Now it works with CUTLASS 2.10.0, 3.0.0 and 3.1.0, as well as with pre-3.1.0 version that is pinned by PyTorch source tree.  Would it be possible to re-try the internal Meta build?

izaitsevfb commented:
@izaitsevfb has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46453477).

cpuhrsch commented:
@pytorchbot merge

alexsamardzic commented:
@pytorchbot rebase

alexsamardzic committed:
Update on "Implement adding bias vector into structured sparse linear operator"




cc nikitaved pearu cpuhrsch amjames bhosmer albanD

Differential Revision: [D46453477](https://our.internmc.facebook.com/intern/diff/D46453477)

[ghstack-poisoned]

malfet commented:
@pytorchbot merge
